Separation of shim/enclave from chaincode app

bvavala opened the pull request:
Signed-off-by: Bruno Vavala <bruno.vavala@intel.com>

mbrandenburger commented on Makefile:
```diff
@@ -1,11 +1,12 @@
 include $(TOP)/build.mk
 
-SUB_DIRS = utils/fabric-ccenv-sgx ercc ecc_enclave ecc tlcc_enclave tlcc integration
+SUB_DIRS = utils/fabric-ccenv-sgx ercc ecc_enclave ecc tlcc_enclave tlcc private-chaincode-apps integration
```
I would rename `private-chaincode-apps` to just `examples`

mbrandenburger commented on private_chaincode_apps/CMakeVariables.txt:
```diff
@@ -0,0 +1,20 @@
+# Copyright 2019 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
```
please use short SPDX license header

bvavala replied:
removed

mbrandenburger commented on private_chaincode_apps/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.5.1)
+project(FPC_APPS)
```
I have the feeling that FPC_APPS is not the right name here. this folder will just contain a chaincode, which may be part of a "bigger" blockchain application. I still like `examples`

bvavala replied:
removed

mbrandenburger commented on private_chaincode_apps/Makefile:
```diff
@@ -0,0 +1,22 @@
+TOP = ..
+include $(TOP)/build.mk
+
+BUILD_DIR := _build
```
does it make sense to have a makefile here? As a developer I would assume that each example chaincode (echo, auction,...) is a self-contained project, maybe similar as the SampleCode in the linux-sgx repo. (https://github.com/intel/linux-sgx/tree/master/SampleCode)

bvavala replied:
yes, just to loop through the examples.
the cmake has been removed

mbrandenburger commented on private_chaincode_apps/echo/echo_cc.cpp:
```diff
@@ -0,0 +1,261 @@
+
+    std::vector<std::string> argss;
+    // parse json args
+    unmarshal_args(argss, args);
```
when I see this, I actually think the shim should provide the args already as array instead of unmarshalling here. I add a note in the shim refactoring issue #83

bvavala replied:
echo has to be cleaned up in a later PR

mbrandenburger commented on private_chaincode_apps/echo/echo_json.h:
```diff
@@ -0,0 +1,28 @@
+    std::string echo_string;
+} echo_t;
+
+int unmarshal(echo_t* echo, const char* json_bytes, uint32_t json_len);
```
maybe json is not even needed in this echo chaincode

bvavala replied:
probably. but it can be enhanced later

bvavala commented:
This PR mainly separates the private chaincode applications from the `ecc` component. Important enhancements include:
* the build of `ecc` produces a boilerplate docker image, which is meant to be later extended with the chaincode app
* the auction chaincode is now an example
* an  echo chaincode has been added to the examples
* the integration test now performs two serial chaincode executions (auction and echo respectively)

This PR must be improved as follows (issues will be created):
* the boilerplate docker image should include the `ecc` binary. This is not possible now because compiling `ecc` requires `mrenclave.go`, which includes the hardcoded MRENCLAVE of the enclave, which is only available once the chaincode application is compiled. Solution: the MRENCLAVE value should be read from a file, which is added to the boilerplate image once the application is compiled.
* the `echo` chaincode must be cleaned up. Also, the write to the ledger is not necessary. Also, the echoed string in the response field of the transaction should be checked.

g2flyer commented on Makefile:
```diff
@@ -1,11 +1,12 @@
 TOP = .
 include $(TOP)/build.mk
 
-SUB_DIRS = utils ercc ecc_enclave ecc tlcc_enclave tlcc integration # docs
```
I actually left the docs there commented out on purpose to explicitly show that we do not make docs now.  But admittedly not an essential thing ...

bvavala replied:
oh, did I remove that?! .. restored!

g2flyer commented on ecc/Dockerfile.boilerplate-ecc:
```diff
@@ -13,7 +13,8 @@ RUN mkdir -p ${CC_LIB_PATH}
 
-copy ecc/${CC_NAME} ${CC_PATH}/chaincode
-copy ecc_enclave/_build/lib/*.so ${CC_LIB_PATH}/
+# TODO: the boiler plate should include the compiled ecc
```
I thought we wanted to outsource the MRENCLAVE into a text file rather than baking it into ecc (and maybe revisit why what's the point of having MRENCLAVE of fpc chaincode used inside ecc in the first place). 
I guess you dropped that as you run out of time?  If so, we should at least add a corresponding issue (and reference it here)?

bvavala replied:
right, this is one of the improvements I have mentioned above.
I will convert these to issues once this is merged

g2flyer commented on ecc/Dockerfile.fpc-app:
```diff
@@ -0,0 +1,14 @@
+RUN test -n "$enclave_so_path" 
+
+copy ${enclave_so_path}/enclave.signed.so ${CC_LIB_PATH}/
+copy ecc/ecc ${CC_PATH}/chaincode
```
Analogous to above boilerplate we might want to add a comment here that this chaincode should eventually go away (and also refer to related issue?)

bvavala replied:
done

g2flyer commented on ecc/Makefile:
```diff
@@ -1,11 +1,13 @@
 include $(TOP)/build.mk
 
-LD_LIBRARY_PATH := LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:$$(pwd)/enclave/lib
+LD_LIBRARY_PATH := LD_LIBRARY_PATH=${LD_LIBRARY_PATH}
```
hmm, if we don't define anything, why not just drop it? Seems rather confusing now (and in general might even be problematic, at least i'm not 100% whether the semantics of LD_LIBRARY_PATH is the same of having it undefined vs defined as empty ...

bvavala replied:
it must not be empty cause the sgx sdk libraries are needed.
however, yes, it can be simply removed (also from hte commands) at this point.
done.

g2flyer commented on ecc/Makefile:
```diff
@@ -15,19 +17,28 @@ VSCC_OUT = ecc-vscc.so
 ERR_MSG = "ecc_enclave build does not exist!"
 
-ecc_enclave:
+ecc_enclave: sym-links
```
I guess all that sym-linking is also due to not-yet removing/outsourcing MRENCLAVE? If so, maybe also add comment as for docker files?

bvavala replied:
done

g2flyer commented on ecc/Makefile:
```diff
@@ -1,11 +1,13 @@
 
-LD_LIBRARY_PATH := LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:$$(pwd)/enclave/lib
+LD_LIBRARY_PATH := LD_LIBRARY_PATH=${LD_LIBRARY_PATH}
+
```
I think you rely on ENCLAVE_SO_PAT for finding fpc-c chaincode (enclave.so).  Maybe also add a define here so all the parameter for the docker build would be at a single place?  (we might also add a commenting header saying that these are the docker build params?)

bvavala replied:
not sure what you mean, but I created an internal variable, now check the value, and clarified what the docker build parameters are

g2flyer replied:
My comment was in the context of the three definitions CC_NAME, NET_ID and PEER_ID which are following at the location i commented:  These three variables define almost but not quite the "api" you would use for 'docker-fpc-app'.  The almost being that you also will have to define that path to the code which i thought is ENCLAVE_SO_PATH.   Hence my suggestion to also add a ?= define or alike for ENCLAVE_SO_PATH to the other three defines and add a comment above the four of them that this defines the 'make docker-fpc-app' api

g2flyer commented on fabric/bin/peer.sh:
```diff
@@ -36,15 +36,19 @@ handle_chaincode_install() {
 		CC_LANG=$2
 		shift;shift
 		;;
+        -e|--enclavesopath)
```
Why add a new parameter? This is really semantically the -p|--path option for language fpc-c?

bvavala replied:
it was not clear at all that -p was there for this.
anyway, fixed

g2flyer commented on examples/echo/echo_json.cpp:
```diff
@@ -0,0 +1,28 @@
+/*
```
I don't see that used. I guess this is the beginning of a utility to help verifying also return results from invocations/queries?   If so, this raises a number of questions:
- couldn't (and shouldn't!) this be generic and hence also should be rather in fabric/bin or alike?
- wouldn't that have been easier scripted in python or alike? (i'd venture, extracting the return result from the peer json can be wrritten in a one-liner python or perl command-line invocation?)
- if we don't complete it, maybe better to leave out of this PR and add to one which does it end-to-end? (With right now just creating a corresponding issue?)

bvavala replied:
the `echo` (as well as the `auction`) cc is not complete. It however performs some operations, and it works now for showing that we can run 2 different cc in serial. That's why it is in this PR, and I will add an issue once merged.

As Marcus pointed out the json machinery here might not be needed, so they might be removed.

g2flyer replied:
oh, i definitely dd not want to pull the echo test from the PR.  My comment purely to the json response processing (in regards of not being echo specific, maybe simpler dealt with a script and better included in this PR only if complete/used (compared to a separate PR which might generalize and also instrument the auction test with checks of return value)

bvavala replied:
the echo test has been cleaned up

bvavala replied:
@g2flyer I think Marcus is waiting for your approval

g2flyer replied:
@bvavala well, i was waiting for you to re-request the review after my initial review. Didn't know when you were done with your changes.  But i assume that is now the case?

bvavala replied:
sorry @g2flyer , there was no button to push for that (I might have done that some days ago).
Anyway, yes, and I have also rebased

g2flyer replied:
actually, there is a button: just click on the "arrow-circle" to the left of the reviewer asking for changes and then s/he will be notified to re-review (that is what marcus has done)

bvavala commented on ecc/Makefile:
```diff
@@ -57,12 +60,24 @@ clean: docker-clean
+	echo "\033[1;33mWARNING: overriding $(CC_NAME) docker image (${DOCKER_IMAGE})\033[0m"
 	$(DOCKER) build -t $(DOCKER_IMAGE) -f Dockerfile ..
 
+docker-boilerplate-ecc:
```
correct

bvavala commented on ecc_enclave/enclave/CMakeLists.txt:
```diff
@@ -50,37 +55,51 @@ include_directories(
-    COMMAND cp ${CMAKE_BINARY_DIR}/enclave.signed.so ${CMAKE_BINARY_DIR}/lib
-    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
-    COMMENT "Create enclave mrenclave")
+    #<<<<<<< Updated upstream
```
this is outdated

bvavala commented on integration/echo_test.sh:
```diff
@@ -0,0 +1,100 @@
+
+CC_ID=ecc
+FPC_APP_NAME=echo
+# TODO: once issue #86 is fixed, change above to ecc_echo_test
```
moved up

bvavala commented on ecc_enclave/enclave/CMakeLists.txt:
```diff
@@ -36,7 +37,11 @@ add_custom_command(
     )
 
-add_library(enclave SHARED ${SOURCE_FILES})
+set(ENCLAVE_BASE_LIB_NAME enclave-base)
```
sounds reasonable, renamed

bvavala commented:
> This looks good to me! I created a hello world example private chaincode to test this. Works now.
> Thanks for this huge piece of work!


Great!
What was wrong with the initial hello-world implementation?

bvavala commented:
> the new version addresses most of my comments, see though a few additional comments. Primarily i'd suggest to already create all follow up issues as otherwise they might get missed with vacations and al.
> 
> Also, there is a format issue in echo_cc.h as 'make' fails as follows ...
> 
> LINT: Running code checks for Cpp/header files... ERROR in format: /home/msteiner/src/poet-et-al/go-trusted-chain-code.v1.4/src/github.com/hyperledger-labs/fabric-private-chaincode/scripts/../examples/echo/echo_cc.h Format check failed. Run './scripts/cpplinter.sh DO_FORMAT' to fix the format.

added 3 issues, and fixed the cpplinter complaint (I still wonder how I missed that since I have tested)

bvavala commented on ecc/Makefile:
```diff
@@ -1,68 +1,83 @@
 
-test:
-	$(LD_LIBRARY_PATH) $(GO) $@ -test.v ./...
+#test:
```
yes, issue #105

bvavala commented on ecc/Makefile:
```diff
@@ -1,68 +1,83 @@
 
-DOCKER_IMAGE ?= $$(docker images | grep -- ${NET_ID}-${PEER_ID}-$(CC_NAME)- | awk '{print $$1;}')
 DOCKER_CONTAINER_ID?=$$(docker ps | grep -- ${NET_ID}-${PEER_ID}-$(CC_NAME)- | awk '{print $$1;}')
+# the following are the required docker build parameters
```
ignored!

bvavala commented on common/enclave/generate_mrenclave.sh:
```diff
@@ -27,7 +28,14 @@ cat $base64_out_name
 echo "Create go file"
 touch $go_name
-echo "package enclave" > $go_name
+echo "/*" >> $go_name
```
kind of, didn't think about that.
Much cleaner now!

mbrandenburger commented:
> Great!
> What was wrong with the initial hello-world implementation?

Cannot really say. Maybe a `make clean` already helped!

BTW. I messed up you PR a bit by squashing and not taking care of a proper commit msg. Sorry :/
